invenio_rdm_records/resources/serializers/csl/__init__.py

```diff
@@ -18,3 +36,89 @@ class CSLJSONSerializer(MarshmallowJSONSerializer):
     def __init__(self, **options):
         """Constructor."""
         super().__init__(schema_cls=CSLJSONSchema, **options)
+
+
+class StringCitationSerializer(CSLJSONSerializer):
+    """CSL Citation Formatter serializer for records.
+
+    In order to produce a formatted citation of a record through citeproc-py,
+    we need a CSL-JSON serialized version of it.
+    """
+
+    _default_style = 'harvard1'
+    """The `citeproc-py` library supports by default the 'harvard1' style."""
+
+    _default_locale = 'en-US'
+    """The `citeproc-py` library supports by default the 'harvard1' style."""
+
+    _user_args = {
+        'style': fields.Str(missing=_default_style),
+        'locale': fields.Str(missing=_default_locale)
+    }
+    """Arguments for the webargs parser."""
+
+    _valid_formats = ('csl', 'bibtex')
+    """Supported formats by citeproc-py."""
+
+    def __init__(self, record_format='csl'):
+        """Initialize the inner record serializer.
+
+        :param record_format: Format that the serializer produces.
+        """
+        super().__init__()
+        self.record_format = record_format
+
+    @classmethod
+    def _get_args(cls, **kwargs):
+        """Parse style and locale.
+
+        Argument location precedence: kwargs > view_args > query
+        """
+        csl_args = {
+            'style': cls._default_style,
+            'locale': cls._default_locale
+        }
+
+        if has_request_context():
+            parser = FlaskParser(locations=('view_args', 'query'))
+            csl_args.update(parser.parse(cls._user_args, request))
+
+        csl_args.update({k: kwargs[k]
+                         for k in ('style', 'locale') if k in kwargs})
+
+        try:
+            csl_args['style'] = get_style_filepath(csl_args['style'].lower())
+        except StyleNotFoundError:
+            if has_request_context():
+                raise RuntimeError(csl_args['style'])
+            raise
+        return csl_args
+
+    def _get_source(self, data):
+        """Get source data object for citeproc-py."""
+        if self.record_format == 'csl':
+            return CiteProcJSON([json.loads(data)])
+        elif self.record_format == 'bibtex':
+            return BibTeX(data)
+
+    def _clean_result(self, text):
+        """Remove double spaces, punctuation and escapes apostrophes."""
+        text = re.sub(r'\s\s+', ' ', text)
+        text = re.sub(r'\.\.+', '.', text)
+        text = text.replace("'", "\\'")
+        return text
+
+    def serialize_object(self, record, **kwargs):
+        """Serialize a single record.
+
+        :param record: Record instance.
+        """
+        serialized_json = super().serialize_object(record)
+        source = CiteProcJSON([json.loads(serialized_json)])
+        style = CitationStylesStyle(validate=False, **self._get_args(**kwargs))
+        bib = CitationStylesBibliography(style, source, formatter.plain)
+
+        citation = Citation([CitationItem(record["id"])])
```

Comment:
Can we access the attribute?
```suggestion
        citation = Citation([CitationItem(record.id)])
```

Reply:
I don't think so, but I will test it again.

Reply:
IIRC, at this point `record` should be a `ResultItem` which has an `id` attribute, see [here](https://github.com/inveniosoftware/invenio-records-resources/blob/master/invenio_records_resources/services/records/results.py#L33)